theano/tensor/nnet/ctc_wrapper.c

```diff
@@ -0,0 +1,292 @@
+#section support_code
+
+typedef struct ctc_context {
+    struct ctcOptions options;
+    void * workspace;
+    int * input_lengths;
+    int * flat_labels;
+    int * label_lengths;
+    PyArrayObject * activations_copy;
+    PyArrayObject * costs;
+    PyArrayObject * gradients;
```

Comment:
In fact, we can ignore these 3 PyArrayObjects:
 - You won't need activations_copy anymore with the new strategy I suggest.
 - We generally don't try to deallocate an output that we have allocated, because this output will be retrieved by Python code of Theano, and will be cleared here. So you don't need to worry about freeing costs and gradients if an error occurs.

So you can just remove them from the context and forget them.